iree/compiler/Dialect/HAL/Transforms/TranslateExecutables.cpp

```diff
@@ -14,105 +14,69 @@
+                         OperationPass<IREE::HAL::ExecutableOp>> {
  public:
   TranslateExecutablesPass()
-      : executableOptions_(getExecutableTargetOptionsFromFlags()) {}
-  explicit TranslateExecutablesPass(ExecutableTargetOptions executableOptions)
+      : executableOptions_(getTargetOptionsFromFlags()) {}
+  explicit TranslateExecutablesPass(TargetOptions executableOptions)
       : executableOptions_(executableOptions) {}
 
   void runOnOperation() override {
-    // Get the target backends we want to translate executables into.
-    SmallVector<std::string, 4> targetBackends;
-    if (executableOptions_.targets.empty()) {
-      for (const auto backend : getExecutableTargetRegistry().keys()) {
-        if (!excludedBackends.count(backend)) {
-          targetBackends.emplace_back(backend);
-        }
-      }
-    } else {
-      for (const auto& targetName : executableOptions_.targets) {
-        auto backendNames = matchExecutableTargetNames(targetName);
-        targetBackends.append(backendNames.begin(), backendNames.end());
-      }
-    }
-    if (targetBackends.empty()) {
-      auto diagnostic = getOperation().emitError();
-      diagnostic
-          << "no target backends available for executable translation; ensure "
-          << "they are linked in and the target options are properly "
-          << "specified. requested = [ ";
-      for (const auto& target : executableOptions_.targets) {
-        diagnostic << "'" << target << "' ";
-      }
-      diagnostic << "], available = [ ";
-      for (const auto& target : getExecutableTargetRegistry().keys()) {
-        diagnostic << "'" << target << "' ";
-      }
-      diagnostic << "]";
-      return signalPassFailure();
-    }
-
-    // Translate all executables to all backends.
-    // When we want heterogenous support we'll do this differently - possibly
-    // even earlier on in the flow dialect (at least, deciding).
-    auto executableOps =
-        llvm::to_vector<32>(getOperation().getOps<IREE::HAL::ExecutableOp>());
-    for (auto executableOp : executableOps) {
-      // Translate for each backend. Variants will be added to the executableOp.
-      for (const auto& targetBackend : targetBackends) {
-        auto targetFn = getExecutableTargetRegistry().find(targetBackend);
-        if (targetFn == getExecutableTargetRegistry().end()) {
-          executableOp.emitError()
-              << "target backend '" << targetBackend << "' unavailable";
-          return signalPassFailure();
-        }
-
-        // Perform translation using the registered translation function.
-        if (failed(targetFn->second(executableOp, executableOptions_))) {
-          executableOp.emitError()
-              << "failed translation to target " << targetBackend;
+    auto executableOp = getOperation();
+    auto targetOps = llvm::to_vector<4>(
+        executableOp.getBlock().getOps<IREE::HAL::ExecutableTargetOp>());
+    for (auto targetOp : targetOps) {
```

Comment:
That's the goal, however today the pass system cannot use any information about the ops to determine whether to run. So nesting on ExecutableTargetOp for the SPIR-V linalg path would also run the nested xla path would also run vmla/etc.

Mehdi and River have a Q2 OKR to enable dynamic pass execution so that we can nest on ExecutableTargetOp *and* a function that checks attributes (like whether this target executable is for SPIR-V, whether it's using the linalg path, etc). At that point I'll be able to remove this pass entirely and have the buildTransformationPipeline in the TargetBackend just add that dynamic pass registration itself. Same with serialization and linking.